### PR TITLE
[ODS-4861] Swagger endpoint in Docker does not work with the virtual path of docs

### PR DESCRIPTION
**Pull request by jamessmckay**

### CODE REVIEW

File: Application/EdFi.Ods.SwaggerUI/ConfigureDefaults.cs [hunk 1]
```diff
@@ -1,39 +0,0 @@
...
-
-            options.DocExpansion(DocExpansion.None);
-            options.DisplayRequestDuration();
-            options.ShowExtensions();
-            options.EnableValidator();
-
-            options.InjectStylesheet(
-                string.IsNullOrWhiteSpace(options.RoutePrefix)
-                    ? "./swagger.css"
-                    : "../swagger.css");
```
Comment (blmeyers) [inline Application/EdFi.Ods.SwaggerUI/ConfigureDefaults.cs:36]: This is probably the cause of everything being pre-expanded.  Might want to double check on some of these other defaults we were previously setting to make sure we're not changing how it works unintentionally.

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 2]
```diff
@@ -3,44 +3,64 @@
...
+                                                   & ForwardedHeaders.XForwardedHost
+                                                   & ForwardedHeaders.XForwardedProto;
+                    });
+            }
         }
 
-        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<Startup> logger)
         {
+            string routePrefix = Configuration.GetValue("SwaggerUIOptions:RoutePrefix", "swagger");
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:54]: Ok will reset the key back to RoutePrefix, and set it as the path base.

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 3]
```diff
@@ -69,9 +96,22 @@ void AppSettingsDelegate(IApplicationBuilder app)
...
+                options =>
+                {
+                    Configuration.Bind("SwaggerUIOptions", options);
+
+                    options.InjectStylesheet(swaggerStyleSheetPath);
+
+                    options.IndexStream = ()
+                        => GetType().Assembly.GetManifestResourceStream("EdFi.Ods.SwaggerUI.Resources.Swashbuckle_index.html");
+
+                    options.RoutePrefix = routePrefix;
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:109]: removed

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 4]
```diff
@@ -52,12 +72,19 @@ void AppSettingsDelegate(IApplicationBuilder app)
...
                                 }));
                     });
             }
 
+            if (!string.IsNullOrEmpty(_pathBase))
+            {
+                app.UsePathBase(_pathBase);
+            }
+
+            app.UseForwardedHeaders();
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:86]: done

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 5]
```diff
@@ -69,9 +96,22 @@ void AppSettingsDelegate(IApplicationBuilder app)
 
             app.Map("/appSettings.json", AppSettingsDelegate);
 
-            app.UseEdFiSwaggerUI();
+            app.UseSwaggerUI(
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:99]: done

File: Application/EdFi.Ods.SwaggerUI/ConfigureDefaults.cs (hunk 1 above)
Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/ConfigureDefaults.cs:36]: fixed

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 6]
```diff
@@ -67,11 +77,56 @@ void AppSettingsDelegate(IApplicationBuilder app)
...
+                    options.EnableValidator();
+                    options.InjectStylesheet(swaggerStyleSheetPath);
+
+                    options.IndexStream = ()
+                        => GetType().Assembly.GetManifestResourceStream("EdFi.Ods.SwaggerUI.Resources.Swashbuckle_index.html");
+
+                    options.ConfigObject.AdditionalItems["WebApiVersionUrl"] = webApiUrl;
+
+                    // default behavior requires "swagger"
+                    options.RoutePrefix = _routePrefix ?? "swagger";
```
Comment (tbmeador-dlp) [inline Application/EdFi.Ods.SwaggerUI/Startup.cs:125]: I think we actually should have used `SwaggerUIOptions:RoutePrefix` originally as it is a part of the `SwaggerUIOptions` class. If we set it this way can we can support both options? RoutePrefix already defaults to "swagger".
```suggestion
                    options.RoutePrefix = _routePrefix ?? options.RoutePrefix;
```

Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:125]: I tried this originally, and this does not work as the bind statement changes the value to null, (unless we define the value). The original code also always set the value to correct this problem.

Comment (jamessmckay) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:125]: I agree that using swaggeroptions:routeprefix is the correct solution over the original implementation of routeprefix. The question becomes, should we make this change, and should we adjust octopus accordingly?

File: Application/EdFi.Ods.SwaggerUI/Startup.cs [hunk 7]
```diff
@@ -3,60 +3,70 @@
...
+                                                   & ForwardedHeaders.XForwardedHost
+                                                   & ForwardedHeaders.XForwardedProto;
                     });
             }
+        }
+
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<Startup> logger)
+        {
+            string webApiUrl = Configuration.GetValue("WebApiVersionUrl", string.Empty);
+            string swaggerStyleSheetPath = string.IsNullOrWhiteSpace(_routePrefix) ? "../swagger.css" : "./swagger.css";
```
[the remark this answers is not in the document]
Comment (blmeyers) [reply Application/EdFi.Ods.SwaggerUI/Startup.cs:58]: Note the original code that was doing these changes happened after we had defaulted in swagger, so it worked fine.  If we move the defaulting to the constructor for route prefix this code may be fine.

**Review by blmeyers — CHANGES_REQUESTED** [review]

**Review by blmeyers — APPROVED** [review]

**Pull request merged**